heat/fft/fft.py

```diff
@@ -0,0 +1,787 @@
+"""Provides a collection of Discrete Fast Fourier Transforms (DFFT) and their inverses."""
+
+import torch
+
+from ..core.communication import MPI
+from ..core.dndarray import DNDarray
+from ..core.stride_tricks import sanitize_axis
+from ..core.types import promote_types, heat_type_of
+from ..core.factories import array, zeros
+
+from typing import Type, Union, Tuple, Any, Iterable, Optional
+
+__all__ = [
+    "fft",
+    "fft2",
+    "fftn",
+    "hfft",
+    "hfft2",
+    "hfftn",
+    "ifft",
+    "ifft2",
+    "ifftn",
+    "ihfft",
+    "ihfft2",
+    "ihfftn",
+    "irfft",
+    "irfft2",
+    "irfftn",
+    "rfft",
+    "rfft2",
+    "rfftn",
+    # "fftfreq",
+    # "rfftfreq",
+    # "fftshift",
+    # "ifftshift",
+]
+
+
+def __fft_op(x: DNDarray, fft_op: callable, **kwargs) -> DNDarray:
+    """
+    Helper function for 1-dimensional FFT.
+    """
+    try:
+        local_x = x.larray
+    except AttributeError:
+        raise TypeError(f"x must be a DNDarray, is {type(x)}")
+    original_split = x.split
+
+    # sanitize kwargs
+    axis = kwargs.get("axis")
+    try:
+        axis = sanitize_axis(x.gshape, axis)
+    except ValueError as e:
+        raise IndexError(e)
+    if isinstance(axis, tuple) and len(axis) > 1:
+        raise TypeError(f"axis must be an integer, got {axis}")
+    n = kwargs.get("n", None)
+    if n is None:
+        n = x.shape[axis]
+    norm = kwargs.get("norm", None)
+
+    # calculate output shape:
+    # if operation requires real input, output size of last transformed dimension is the Nyquist frequency
+    output_shape = list(x.shape)
+    real_to_generic_fft_ops = {
+        torch.fft.rfft: torch.fft.fft,
+        torch.fft.ihfft: torch.fft.ifft,
+    }
+    real_op = fft_op in real_to_generic_fft_ops
+    if real_op:
+        nyquist_freq = n // 2 + 1
+        output_shape[axis] = nyquist_freq
+    else:
+        output_shape[axis] = n
+    print("DEBUGGING: n, output_shape = ", n, output_shape)
```

Comment:
here, and at some other places debuggin-info gets printed; we should remove that from the code before merging

Reply:
You're right, removed